### ISSUE TITLE
Filtering

### ISSUE THREAD

**Issue by dominikh** [dominikh/gotraceui#6]
Implement customizable filtering of goroutines and spans. 

At a minimum we want to filter goroutines based on

- the function name at the bottom of their stack
- their ID
- whether they contain any spans matching span filters
- interactions; given a primary goroutine, show goroutines that interact with it, i.e. unblock it or get unblocked by it
- their user regions, tasks, and logs
- the proc they ran on

At a minimum we want to filter spans based on

- their state
- their tags
- their duration
- their start or end time
- whether they overlap with GC or STW
- their user regions, tasks, and logs

**Comment by prateek** [comment]
Hey I tried using the tool, and found this is the biggest thing holding me back. Can i help with this at all with this?

**Comment by dominikh** [comment]
> Can i help with this at all with this?

Thank you for the offer, but my bandwidth is already taken up by working on Gotraceui, I don't think that external contributions would speed things up.

**Comment by prateek** [comment]
Fair enough, I appreciate the quick response nonetheless :) 

For my immediate needs - I was able to use your Timelines view (`g`) to isolate to the call paths and find the stacktraces manually after. Minor suggestions (happy to file a separate PR if you'd like) that I think would drive the adoption/usage of this tool:

- Expose a keyboard shortcuts overlay triggered by `?` (akin to what Gmail/Vimium/etc do) - just highlighting the different keys and what they do. I have read the manual but still managed to not find the `g` shortcut on my first pass. 
- The PDF manual you publish is spectacular, but it is dense. If you had a blog/video/wiki showing the workflow of how you use this tool, it would make it a lot easier for people to Grok. (e.g. something akin to this [post](https://about.sourcegraph.com/blog/go/an-introduction-to-go-tool-trace-rhys-hiltner) would lower the barrier for entry)

Anyhow, thank you for the hard work on this tool. From 20m reading the code/PRs/references, it's quite clear you've spent a lot of effort into this, and it shows :)

**Comment by dominikh** [comment]
> Expose a keyboard shortcuts overlay triggered by ? (akin to what Gmail/Vimium/etc do) - just highlighting the different keys and what they do. I have read the manual but still managed to not find the g shortcut on my first pass.

We'll probably end up doing that, yeah. Currently, the (other) way to discover shortcuts is by looking through the menus. However, those are currently missing an entry for `g`…

> If you had a blog/video/wiki showing the workflow of how you use this tool

It is my hope that the community will provide these eventually. Different people will use the tool very differently, depending on the nature of their code, and I couldn't possibly cover all of those.

At the same time, I think it is too early to create such material. As the tool improves, workflows will change, things in the UI will move around, etc. This will quickly render guides outdated.

> lower the barrier for entry

While ultimately that's the obvious goal, I think we're still in the early phase of development, making you an early adopter. That is to say, there will be some pain.


> The PDF manual you publish is spectacular, but it is dense.
> Anyhow, thank you for the hard work on this tool.

Thank you!
